lstchain/reco/reconstructor.py

```diff
@@ -0,0 +1,977 @@
+import inspect
+import logging
+
+from iminuit import Minuit
+import numpy as np
+import numexpr as ne
+import matplotlib.pyplot as plt
+from copy import copy
+import astropy.units as u
+
+from lstchain.data.normalised_pulse_template import NormalizedPulseTemplate
+
+from ctapipe.core import Component
+from ctapipe.core.traits import Bool, Float, Int, Path
+from lstchain.calib.camera.pixel_threshold_estimation import get_bias_and_std
+from lstchain.io.lstcontainers import DL1LikelihoodParametersContainer
+from lstchain.image.pdf import log_gaussian, log_asygaussian2d
+from lstchain.visualization.camera import display_array_camera
+
+logger = logging.getLogger(__name__)
+
+
+class TimeWaveformFitter(Component):
+    sigma_s = Float(1, help='', allow_none=False).tag(config=True)
+    crosstalk = Float(0, help='', allow_none=False).tag(config=True)
+    sigma_space = Float(4, help='', allow_none=False).tag(config=True)
+    sigma_time = Float(3, help='', allow_none=False).tag(config=True)
+    time_before_shower = Float(10, help='', allow_none=False).tag(config=True)
+    time_after_shower = Float(20, help='', allow_none=False).tag(config=True)
+    use_weight = Bool(False, help='', allow_none=False).tag(config=True)
+    no_asymmetry = Bool(False, help='', allow_none=False).tag(config=True)
+    use_interleaved = Path(None, help='', allow_none=True).tag(config=True)
+    telescope_id = Int(1, help='', allow_none=False).tag(config=True)
```

Comment:
Why do you require a `telescope_id` here?

Reply:
It is used to select a couple of information (select geometry for the pixel positions, the sampling frequency). Later when going to stereo, It may be useful to select a subarray for a full run. And a second event wise list of id would also be needed. 
I assumed it was faster to save, e.g. the pixels position before masking (mask per event) and the sampling frequency instead of using the telescope id at each event. I can change this if I am wrong.

Reply:
Since both are just a dictionary lookup, it won't be much faster to do something like

```
self.sampling_frequency
```
instead of
```
self.subarray.tel[tel_id].camera.readout.sampling_frequency 
```

especially considering, that you are performing a pretty expensive likelihood computation, this will be completely negligible. So yes, please do not limit to setting the telescope id via the config. 

If there is something really expensive to be done per telescope, it can be cached.

Reply:
So you are saying to pass and additional argument and to move all this to __call__ ? :
        self.geometry = self.subarray.tel[self.telescope_id].camera.geometry
        self.pix_x = self.geometry.pix_x.to_value(u.m)
        self.pix_y = self.geometry.pix_y.to_value(u.m)
        self.pix_area = self.geometry.pix_area.to_value(u.m ** 2)
        self.r_max = np.sqrt(self.pix_x ** 2 + self.pix_y ** 2).max()
        self.focal_length = subarray.tel[self.telescope_id].optics.equivalent_focal_length
        readout = self.subarray.tel[self.telescope_id].camera.readout
        sampling_rate = readout.sampling_rate.to_value(u.GHz)
        self.dt = (1.0 / sampling_rate)
and this (i forgot to pass the telescope_id here, I will change it)
        self.template = NormalizedPulseTemplate.load_from_eventsource(subarray.tel[1].camera.readout)
        self.template_time_of_max = self.template.compute_time_of_max()

It seems like a waste of resources and I don't understand what would be the advantage.